onnxruntime/core/providers/cpu/tensor/unique.cc

```diff
@@ -0,0 +1,354 @@
+            OpSchema::Optional)
+        .Output(
+            2,
+            "inverse_indices",
+            "A 1-D INT64 tensor "
+            "containing, for elements of 'X', its corresponding indices in 'Y'. "
+            "When 'axis' is provided, it contains indices to subtensors in output 'Y' on the 'axis'. "
+            "When 'axis' is not provided, it contains indices to values in output 'Y'. ",
+            "tensor(int64)",
+            OpSchema::Optional)
+        .Output(
+            3,
+            "counts",
+            "A 1-D INT64 tensor containing "
+            "the count of each element "
+            "of 'Y' in input 'X'",
+            "tensor(int64)",
+            OpSchema::Optional)
+        .TypeConstraint(
+            "T",
+            OpSchema::all_tensor_types(),
+            "Input can be of any tensor type.")    
+*/
+ONNX_CPU_OPERATOR_KERNEL(
+    Unique,
+    11,
+    KernelDefBuilder().TypeConstraint("T", DataTypeImpl::AllTensorTypes()),
+    Unique);
+
+Status Unique::Compute(OpKernelContext* context) const {
+  const Tensor& input = *context->Input<Tensor>(0);
+
+  Status status;
+  auto data_type = input.DataType();
+
+  DispatchOnTensorTypeWithReturn(data_type, status, ComputeImpl, *context);
+
+  return status;
+}
+
+// class to represent a subtensor along a given axis for a single entry on that axis
+template <typename T>
+class Subtensor {
+ public:
+  // Create Subtensor for entry 'idx' on axis 'axis'
+  // n_axis is the number of entries for 'axis' is the original data.
+  //   e.g. if original shape was [4, 2] and axis is 1, n_axis == 2.
+  // subtensor_shape is the shape for the subtensor. the dimension value for the 'axis' dimension in it will be 1.
+  Subtensor(const gsl::span<const T>& data, const TensorShape& subtensor_shape,
+            int64_t axis, int64_t n_axis, int64_t idx) {
+    // rows and columns for the slice along axis, flattened to 2D by merging the dimensions before and after the axis
+    int64_t columns = subtensor_shape.SizeFromDimension(axis);
+    int64_t rows = subtensor_shape.SizeToDimension(axis);
+    items_.reserve(rows * columns);
+    size_t cur_data = idx * columns;  // offset into data for first row of slice
+
+    for (int r = 0; r < rows; ++r) {
+      for (int c = 0; c < columns; ++c) {
+        // TODO: could copy blocks instead of individual items for non std::string types
+        items_.push_back(data[cur_data + c]);
+      }
+
+      cur_data += columns * n_axis;
+    }
+  }
+
+  bool operator<(const Subtensor& rhs) const {
+    return items_ < rhs.items_;
+  }
+
+  const std::vector<T>& GetItems() const { return items_; }
+
+ private:
+  // TODO: Simple copy for now. std::string would be better as std::reference_wrapper<std::string>
+  std::vector<T> items_;
+};
+
+template <typename T>
+static void CreateFlattenedOutput(OpKernelContext& context,
+                                  const std::map<const T, int64_t>& offsets,         // sorted:unsorted idx
```

Comment:
Clarification: so I guess the comment means - "sorted data : unsorted indices". Correct ?

Reply:
Would 'map sorted key to unsorted idx' be better?

Reply:
yes